saleor/static/dashboard-next/discounts/components/DiscountCategories/DiscountCategories.tsx

```diff
@@ -0,0 +1,148 @@
+import { ListProps } from "../../../types";
+import { SaleDetails_sale } from "../../types/SaleDetails";
+import { VoucherDetails_voucher } from "../../types/VoucherDetails";
+
+export interface DiscountCategoriesProps extends ListProps {
+  discount: SaleDetails_sale | VoucherDetails_voucher;
+  onCategoryAssign: () => void;
+  onCategoryUnassign: (id: string) => void;
+}
+
+const styles = (theme: Theme) =>
+  createStyles({
+    iconCell: {
+      "&:last-child": {
+        paddingRight: 0
+      },
+      width: 48 + theme.spacing.unit / 2
+    },
+    tableRow: {
+      cursor: "pointer"
+    },
+    textRight: {
+      textAlign: "right"
+    },
+    wideColumn: {
+      width: "60%"
+    }
+  });
+const DiscountCategories = withStyles(styles, {
+  name: "DiscountCategories"
+})(
+  ({
+    discount: sale,
+    classes,
+    disabled,
+    pageInfo,
+    onCategoryAssign,
+    onCategoryUnassign,
+    onRowClick,
+    onPreviousPage,
+    onNextPage
+  }: DiscountCategoriesProps & WithStyles<typeof styles>) => (
+    <Card>
+      <CardTitle
+        title={i18n.t("Categories assigned to {{ saleName }}", {
+          saleName: maybe(() => sale.name)
+        })}
+        toolbar={
+          <Button variant="flat" color="secondary" onClick={onCategoryAssign}>
+            {i18n.t("Assign categories")}
+          </Button>
+        }
+      />
+      <Table>
+        <TableHead>
+          <TableRow>
+            <TableCell className={classes.wideColumn}>
+              {i18n.t("Category name")}
+            </TableCell>
+            <TableCell className={classes.textRight}>
+              {i18n.t("Products")}
+            </TableCell>
+            <TableCell />
+          </TableRow>
+        </TableHead>
+        <TableFooter>
+          <TableRow>
+            <TablePagination
+              colSpan={3}
+              hasNextPage={pageInfo && !disabled ? pageInfo.hasNextPage : false}
+              onNextPage={onNextPage}
+              hasPreviousPage={
+                pageInfo && !disabled ? pageInfo.hasPreviousPage : false
+              }
+              onPreviousPage={onPreviousPage}
+            />
+          </TableRow>
+        </TableFooter>
+        <TableBody>
+          {renderCollection(
+            maybe(() => sale.categories.edges.map(edge => edge.node)),
+            category => (
+              <TableRow
+                hover={!!category}
+                key={category ? category.id : "skeleton"}
+                onClick={category && onRowClick(category.id)}
+                className={classes.tableRow}
+              >
+                <TableCell>
+                  {maybe<React.ReactNode>(() => category.name, <Skeleton />)}
+                </TableCell>
+                <TableCell className={classes.textRight}>
+                  {maybe<React.ReactNode>(
+                    () => category.products.totalCount,
+                    <Skeleton />
+                  )}
```

Comment:
U should really extract this logic into a separate helper component instead providing this every time. I saw this pattern repeated a lot in dashboard code.
`<ValOrSkeleton check={category.products.totalCount} />`
Name it whatever u want.

Reply:
Good idea, I left a ticket for this #3705